expose noninterpolative scaling in direct mode and ncplayer, unfuck ncvisual_inflate()

@dankamongmen  I've just added NCVISUAL_OPTION_NOINTERPOLATE to the rust bindings, and it works. Thank you for finishing it so quickly! I just updated the pixel-cell example to include it.

In that example I compare interpolative scaling, non-interpolative scaling, inflate and resize... And I also checked that ncvisual_resize always uses interpolation, even when using the NCVISUAL_OPTION_NOINTERPOLATE flag. That made me think.

1. Have you thought about maybe removing ncvisual_inflate and/or using ncvisual_resize in the same way scaling works now, for consistency, if not as a substitution as an alternative? 

2. And another issue is that downsizing to less than the original size makes the subsequent ncvisual_render fail... I don't remember if that's already in the roadmap or a nonfix...

and, as expected:

```rust
	   Compiling libnotcurses-sys v2.3.2 (/drone/src/rust)
188s
3287	error[E0061]: this function takes 3 arguments but 6 arguments were supplied
194s
3288	   --> src/visual/methods.rs:523:17
194s
3289	    |
194s
3290	523 |                 crate::ncdirectf_render(ncd, self, blitter, scale, max_y as i32, max_x as i32)
194s
3291	    |                 ^^^^^^^^^^^^^^^^^^^^^^^ ---  ----  -------  -----  ------------  ------------ supplied 6 arguments
194s
3292	    |                 |
194s
3293	    |                 expected 3 arguments
194s
3294	    | 
194s
3295	   ::: /drone/src/rust/target/debug/build/libnotcurses-sys-8d43744d70e4f6e0/out/bindings.rs:3:240694
194s
3296	    |
194s
3297	3   | ..."] pub fn ncdirectf_render (n : * mut ncdirect , frame : * mut ncdirectf , vopts : * mut ncvisual_options) -> * mut ncdirectv ; } exte...
194s
3298	    |       ---------------------------------------------------------------------------------------------------------------------------- defined here
```

--- a/NEWS.md
+++ b/NEWS.md
@@ -11,6 +11,11 @@ rearrangements of Notcurses.
   * `ncvisual_inflate()` has been rewritten as a wrapper around the new
     function `ncvisual_resize_noninterpolative()`, and deprecated. It will be
     removed for ABI3. Godspeed, `ncvisual_inflate()`; we hardly knew ye.
+  * `ncdirect_renderf()` has been changed to accept a `ncvisual_options`,
+    replacing and extending its four final arguments. Sorry about the breakage
+    here, but `ncdirect_renderf()` was introduced pretty recently (2.3.1).
+    As a result, `ncdirect_renderf()` and `ncdirect_stream()` now honor
+    `NCVISUAL_OPTION_BLEND` and `NCVISUAL_OPTION_NOINTERPOLATE`.
 
 * 2.3.2 (2021-06-03)
   * Fixed a bug affecting certain scalings of `ncvisual` objects created from

--- a/NEWS.md
+++ b/NEWS.md
@@ -16,6 +16,7 @@ rearrangements of Notcurses.
     here, but `ncdirect_renderf()` was introduced pretty recently (2.3.1).
     As a result, `ncdirect_renderf()` and `ncdirect_stream()` now honor
     `NCVISUAL_OPTION_BLEND` and `NCVISUAL_OPTION_NOINTERPOLATE`.
+  * `ncplayer` now accepts `-n` to force non-interpolative scaling.
 
 * 2.3.2 (2021-06-03)
   * Fixed a bug affecting certain scalings of `ncvisual` objects created from

--- a/doc/man/man1/ncplayer.1.md
+++ b/doc/man/man1/ncplayer.1.md
@@ -8,7 +8,7 @@ ncplayer - Render images and video to a terminal
 
 # SYNOPSIS
 
-**ncplayer** [**-h**] [**-V**] [**-q**] [**-d** ***delaymult***] [**-l** ***loglevel***] [**-b** ***blitter***] [**-s** ***scalemode***] [**-k**] [**-L**] [**-t** ***seconds***] [**-a**] files
+**ncplayer** [**-h**] [**-V**] [**-q**] [**-d** ***delaymult***] [**-l** ***loglevel***] [**-b** ***blitter***] [**-s** ***scalemode***] [**-k**] [**-L**] [**-t** ***seconds***] [**-n**] [**-a**] files
 
 # DESCRIPTION
 
@@ -42,6 +42,8 @@ be any non-negative number.
 **-q**: Print neither frame/timing information along the top of the screen, nor the output summary on exit.
 
 **-a**: Treat color 0x000000 as if it were transparent.
+
+**-n**: Use non-interpolative scaling. The result is usually less pleasing to the eye, but it doesn't introduce new colors.
 
 **-V**: Print the program name and version, and exit with success.
 

--- a/doc/man/man3/notcurses_direct.3.md
+++ b/doc/man/man3/notcurses_direct.3.md
@@ -116,7 +116,7 @@ typedef struct ncvgeom {
 
 **void ncdirectf_free(struct ncdirectf* ***frame***);**
 
-**ncdirectv* ncdirectf_render(struct ncdirect* ***n***, struct ncdirectf* ***frame***, struct ncvisual_options ***vopts***);**
+**ncdirectv* ncdirectf_render(struct ncdirect* ***n***, struct ncdirectf* ***frame***, const struct ncvisual_options ***vopts***);**
 
 **int ncdirectf_geom(struct ncdirect* ***n***, struct ncdirectf* ***frame***, ncblitter_e* ***blitter***, ncscale_e ***scale***, int ***maxy***, int ***maxx***, ncvgeom* ***geom***);**
 

--- a/doc/man/man3/notcurses_direct.3.md
+++ b/doc/man/man3/notcurses_direct.3.md
@@ -116,7 +116,7 @@ typedef struct ncvgeom {
 
 **void ncdirectf_free(struct ncdirectf* ***frame***);**
 
-**ncdirectv* ncdirectf_render(struct ncdirect* ***n***, struct ncdirectf* ***frame***, ncblitter_e ***blitter***, ncscale_e ***scale***, int ***maxy***, int ***maxx***);**
+**ncdirectv* ncdirectf_render(struct ncdirect* ***n***, struct ncdirectf* ***frame***, struct ncvisual_options ***vopts***);**
 
 **int ncdirectf_geom(struct ncdirect* ***n***, struct ncdirectf* ***frame***, ncblitter_e* ***blitter***, ncscale_e ***scale***, int ***maxy***, int ***maxx***, ncvgeom* ***geom***);**
 

--- a/include/notcurses/direct.h
+++ b/include/notcurses/direct.h
@@ -389,7 +389,7 @@ API void ncdirectf_free(ncdirectf* frame);
 // loaded. A loaded frame may be rendered in different ways before it is
 // destroyed.
 API ALLOC ncdirectv* ncdirectf_render(struct ncdirect* n, ncdirectf* frame,
-                                      struct ncvisual_options* vopts)
+                                      const struct ncvisual_options* vopts)
   __attribute__ ((nonnull (1, 2)));
 
 // Having loaded the frame 'frame', get the geometry of a potential render.

--- a/include/notcurses/direct.h
+++ b/include/notcurses/direct.h
@@ -389,8 +389,7 @@ API void ncdirectf_free(ncdirectf* frame);
 // loaded. A loaded frame may be rendered in different ways before it is
 // destroyed.
 API ALLOC ncdirectv* ncdirectf_render(struct ncdirect* n, ncdirectf* frame,
-                                      ncblitter_e blitter, ncscale_e scale,
-                                      int maxy, int maxx)
+                                      struct ncvisual_options* vopts)
   __attribute__ ((nonnull (1, 2)));
 
 // Having loaded the frame 'frame', get the geometry of a potential render.

--- a/rust/examples/issue-1716.rs
+++ b/rust/examples/issue-1716.rs
@@ -19,7 +19,9 @@ fn main() {
     }
 
     let vframe1 = NcDirectF::from_bgra(&buffer, H, W * 4, W).expect("couldn’t create visual");
-    let v = vframe1.ncdirectf_render(nc, NCBLIT_PIXEL, NCSCALE_NONE, 0, 0).expect("failed to render image to sixels");
+    let mut voptions1 = NcVisualOptions::without_plane(0, 0, 0, 0, H, W, NCBLIT_PIXEL, 0, 0);
+    let v = vframe1.ncdirectf_render(nc, &mut voptions1)
+        .expect("failed to render image to sixels");
     nc.raster_frame(v, NCALIGN_LEFT).expect("failed to print sixels");
     vframe1.ncdirectf_free();
     nc.stop().expect("failed to destroy ncdirect context");

--- a/rust/src/visual/methods.rs
+++ b/rust/src/visual/methods.rs
@@ -513,14 +513,11 @@ impl NcDirectF {
     pub fn ncdirectf_render(
         &mut self,
         ncd: &mut NcDirect,
-        blitter: NcBlitter,
-        scale: NcScale,
-        max_y: NcDim,
-        max_x: NcDim,
+        options: &mut NcVisualOptions,
     ) -> NcResult<&mut NcDirectV> {
         error_ref_mut![
             unsafe {
-                crate::ncdirectf_render(ncd, self, blitter, scale, max_y as i32, max_x as i32)
+                crate::ncdirectf_render(ncd, self, options)
             },
             "NcVisual.render()"
         ]

--- a/src/lib/direct.c
+++ b/src/lib/direct.c
@@ -626,6 +626,7 @@ ncdirect_render_visual(ncdirect* n, ncvisual* ncv, struct ncvisual_options* vopt
     return NULL;
   }
   blitterargs bargs = {};
+  bargs.flags = vopts->flags;
   if(vopts->flags & NCVISUAL_OPTION_ADDALPHA){
     bargs.transcolor = vopts->transcolor | 0x1000000ull;
   }

--- a/src/lib/direct.c
+++ b/src/lib/direct.c
@@ -556,7 +556,9 @@ ncdirect_render_visual(ncdirect* n, ncvisual* ncv, struct ncvisual_options* vopt
   }
 //fprintf(stderr, "OUR DATA: %p rows/cols: %d/%d outsize: %d/%d %d/%d\n", ncv->data, ncv->pixy, ncv->pixx, dimy, dimx, ymax, xmax);
 //fprintf(stderr, "render %d/%d to scaling: %d\n", ncv->pixy, ncv->pixx, vopts->scaling);
-  const struct blitset* bset = rgba_blitter_low(&n->tcache, vopts->scaling, true, vopts->blitter);
+  const struct blitset* bset = rgba_blitter_low(&n->tcache, vopts->scaling,
+                                                !(vopts->flags & NCVISUAL_OPTION_NODEGRADE),
+                                                vopts->blitter);
   if(!bset){
     return NULL;
   }
@@ -647,14 +649,31 @@ ncdirect_render_visual(ncdirect* n, ncvisual* ncv, struct ncvisual_options* vopt
 ncdirectv* ncdirect_render_frame(ncdirect* n, const char* file,
                                  ncblitter_e blitfxn, ncscale_e scale,
                                  int ymax, int xmax){
+  if(ymax < 0 || xmax < 0){
+    return NULL;
+  }
   ncdirectf* ncv = ncdirectf_from_file(n, file);
   if(ncv == NULL){
     return NULL;
   }
   struct ncvisual_options vopts = {};
-  vopts.blitter = blitfxn;
+  const struct blitset* bset = rgba_blitter_low(&n->tcache, scale, true, blitfxn);
+  if(!bset){
+    return NULL;
+  }
+  vopts.blitter = bset->geom;
+  vopts.flags = NCVISUAL_OPTION_NODEGRADE;
   vopts.scaling = scale;
-  // FIXME convery ymax, xmax into leny/lenx, or something
+  if(ymax > 0){
+    if((vopts.leny = ymax * bset->height) > ncv->pixy){
+      vopts.leny = 0;
+    }
+  }
+  if(xmax > 0){
+    if((vopts.lenx = xmax * bset->width) > ncv->pixx){
+      vopts.lenx = 0;
+    }
+  }
   ncdirectv* v = ncdirectf_render(n, ncv, &vopts);
   ncvisual_destroy(ncv);
   return v;

--- a/src/lib/direct.c
+++ b/src/lib/direct.c
@@ -545,7 +545,8 @@ int ncdirect_raster_frame(ncdirect* n, ncdirectv* ncdv, ncalign_e align){
 }
 
 static ncdirectv*
-ncdirect_render_visual(ncdirect* n, ncvisual* ncv, struct ncvisual_options* vopts){
+ncdirect_render_visual(ncdirect* n, ncvisual* ncv,
+                       const struct ncvisual_options* vopts){
   struct ncvisual_options defvopts = {};
   if(!vopts){
     vopts = &defvopts;
@@ -1313,7 +1314,7 @@ void ncdirectf_free(ncdirectf* frame){
   ncvisual_destroy(frame);
 }
 
-ncdirectv* ncdirectf_render(ncdirect* n, ncdirectf* frame, struct ncvisual_options* vopts){
+ncdirectv* ncdirectf_render(ncdirect* n, ncdirectf* frame, const struct ncvisual_options* vopts){
   return ncdirect_render_visual(n, frame, vopts);
 }
 

--- a/src/lib/direct.c
+++ b/src/lib/direct.c
@@ -545,23 +545,27 @@ int ncdirect_raster_frame(ncdirect* n, ncdirectv* ncdv, ncalign_e align){
 }
 
 static ncdirectv*
-ncdirect_render_visual(ncdirect* n, ncvisual* ncv, ncblitter_e blitfxn,
-                       ncscale_e scale, int ymax, int xmax,
-                       uint32_t transcolor){
-  if(ymax < 0 || xmax < 0){
-    fprintf(stderr, "Invalid render geometry %d/%d\n", ymax, xmax);
+ncdirect_render_visual(ncdirect* n, ncvisual* ncv, struct ncvisual_options* vopts){
+  struct ncvisual_options defvopts = {};
+  if(!vopts){
+    vopts = &defvopts;
+  }
+  if(vopts->leny < 0 || vopts->lenx < 0){
+    fprintf(stderr, "Invalid render geometry %d/%d\n", vopts->leny, vopts->lenx);
     return NULL;
   }
-  int dimy = ymax > 0 ? ymax : (ncdirect_dim_y(n) - 1);
-  int dimx = xmax > 0 ? xmax : ncdirect_dim_x(n);
 //fprintf(stderr, "OUR DATA: %p rows/cols: %d/%d outsize: %d/%d %d/%d\n", ncv->data, ncv->pixy, ncv->pixx, dimy, dimx, ymax, xmax);
-//fprintf(stderr, "render %d/%d to scaling: %d\n", ncv->pixy, ncv->pixx, scale);
-  const struct blitset* bset = rgba_blitter_low(&n->tcache, scale, true, blitfxn);
+//fprintf(stderr, "render %d/%d to scaling: %d\n", ncv->pixy, ncv->pixx, vopts->scaling);
+  const struct blitset* bset = rgba_blitter_low(&n->tcache, vopts->scaling, true, vopts->blitter);
   if(!bset){
     return NULL;
   }
+  int ymax = vopts->leny / bset->height;
+  int xmax = vopts->lenx / bset->width;
+  int dimy = vopts->leny > 0 ? ymax : ncdirect_dim_y(n);
+  int dimx = vopts->lenx > 0 ? xmax : ncdirect_dim_x(n);
   int disprows, dispcols, outy;
-  if(scale != NCSCALE_NONE && scale != NCSCALE_NONE_HIRES){
+  if(vopts->scaling != NCSCALE_NONE && vopts->scaling != NCSCALE_NONE_HIRES){
     if(bset->geom != NCBLIT_PIXEL){
       dispcols = dimx * encoding_x_scale(&n->tcache, bset);
       disprows = dimy * encoding_y_scale(&n->tcache, bset);
@@ -569,20 +573,20 @@ ncdirect_render_visual(ncdirect* n, ncvisual* ncv, ncblitter_e blitfxn,
     }else{
       dispcols = dimx * n->tcache.cellpixx;
       disprows = dimy * n->tcache.cellpixy;
-      clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, scale);
+      clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, vopts->scaling);
     }
-    if(scale == NCSCALE_SCALE || scale == NCSCALE_SCALE_HIRES){
+    if(vopts->scaling == NCSCALE_SCALE || vopts->scaling == NCSCALE_SCALE_HIRES){
       scale_visual(ncv, &disprows, &dispcols);
       outy = disprows;
       if(bset->geom == NCBLIT_PIXEL){
-        clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, scale);
+        clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, vopts->scaling);
       }
     }
   }else{
     disprows = ncv->pixy;
     dispcols = ncv->pixx;
     if(bset->geom == NCBLIT_PIXEL){
-      clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, scale);
+      clamp_to_sixelmax(&n->tcache, &disprows, &dispcols, &outy, vopts->scaling);
     }else{
       outy = disprows;
     }
@@ -620,7 +624,9 @@ ncdirect_render_visual(ncdirect* n, ncvisual* ncv, ncblitter_e blitfxn,
     return NULL;
   }
   blitterargs bargs = {};
-  bargs.transcolor = transcolor;
+  if(vopts->flags & NCVISUAL_OPTION_ADDALPHA){
+    bargs.transcolor = vopts->transcolor | 0x1000000ull;
+  }
   if(bset->geom == NCBLIT_PIXEL){
     bargs.u.pixel.celldimx = n->tcache.cellpixx;
     bargs.u.pixel.celldimy = n->tcache.cellpixy;
@@ -645,7 +651,11 @@ ncdirectv* ncdirect_render_frame(ncdirect* n, const char* file,
   if(ncv == NULL){
     return NULL;
   }
-  ncdirectv* v = ncdirectf_render(n, ncv, blitfxn, scale, ymax, xmax);
+  struct ncvisual_options vopts = {};
+  vopts.blitter = blitfxn;
+  vopts.scaling = scale;
+  // FIXME convery ymax, xmax into leny/lenx, or something
+  ncdirectv* v = ncdirectf_render(n, ncv, &vopts);
   ncvisual_destroy(ncv);
   return v;
 }
@@ -1249,10 +1259,7 @@ int ncdirect_stream(ncdirect* n, const char* filename, ncstreamcb streamer,
     if(x > 0){
       ncdirect_cursor_left(n, x);
     }
-    // FIXME what about vopts->beg{yx} and vopts->len{yx}?
-    ncdirectv* v = ncdirect_render_visual(n, ncv, vopts->blitter, vopts->scaling,
-                                          0, 0, (vopts->flags & NCVISUAL_OPTION_ADDALPHA) ?
-                                                 vopts->transcolor | 0x1000000ul : 0);
+    ncdirectv* v = ncdirect_render_visual(n, ncv, vopts);
     if(v == NULL){
       ncvisual_destroy(ncv);
       return -1;
@@ -1286,10 +1293,8 @@ void ncdirectf_free(ncdirectf* frame){
   ncvisual_destroy(frame);
 }
 
-ncdirectv* ncdirectf_render(ncdirect* n, ncdirectf* frame,
-                            ncblitter_e blitter, ncscale_e scale,
-                            int maxy, int maxx){
-  return ncdirect_render_visual(n, frame, blitter, scale, maxy, maxx, 0);
+ncdirectv* ncdirectf_render(ncdirect* n, ncdirectf* frame, struct ncvisual_options* vopts){
+  return ncdirect_render_visual(n, frame, vopts);
 }
 
 int ncdirectf_geom(ncdirect* n, ncdirectf* frame,

--- a/src/player/play.cpp
+++ b/src/player/play.cpp
@@ -19,7 +19,7 @@ static void usage(std::ostream& os, const char* name, int exitcode)
   __attribute__ ((noreturn));
 
 void usage(std::ostream& o, const char* name, int exitcode){
-  o << "usage: " << name << " [ -h ] [ -q ] [ -m margins ] [ -l loglevel ] [ -d mult ] [ -s scaletype ] [ -k ] [ -L ] [ -t seconds ] [ -a ] files" << '\n';
+  o << "usage: " << name << " [ -h ] [ -q ] [ -m margins ] [ -l loglevel ] [ -d mult ] [ -s scaletype ] [ -k ] [ -L ] [ -t seconds ] [ -n ] [ -a ] files" << '\n';
   o << " -h: display help and exit with success\n";
   o << " -V: print program name and version\n";
   o << " -q: be quiet (no frame/timing information along top of screen)\n";
@@ -31,6 +31,7 @@ void usage(std::ostream& o, const char* name, int exitcode){
   o << " -b blitter: one of 'ascii', 'half', 'quad', 'sex', 'braille', or 'pixel'\n";
   o << " -m margins: margin, or 4 comma-separated margins\n";
   o << " -a: replace color 0x000000 with a transparent channel\n";
+  o << " -n: force non-interpolative scaling\n";
   o << " -d mult: non-negative floating point scale for frame time" << std::endl;
   exit(exitcode);
 }
@@ -184,13 +185,14 @@ auto perframe(struct ncvisual* ncv, struct ncvisual_options* vopts,
 // can exit() directly. returns index in argv of first non-option param.
 auto handle_opts(int argc, char** argv, notcurses_options& opts, bool* quiet,
                  float* timescale, ncscale_e* scalemode, ncblitter_e* blitter,
-                 float* displaytime, bool* loop, uint32_t* transcolor)
+                 float* displaytime, bool* loop, bool* noninterp,
+                 uint32_t* transcolor)
                  -> int {
   *timescale = 1.0;
   *scalemode = NCSCALE_STRETCH;
   *displaytime = -1;
   int c;
-  while((c = getopt(argc, argv, "Vhql:d:s:b:t:m:kLa")) != -1){
+  while((c = getopt(argc, argv, "Vhql:d:s:b:t:m:kLan")) != -1){
     switch(c){
       case 'h':
         usage(std::cout, argv[0], EXIT_SUCCESS);
@@ -198,6 +200,13 @@ auto handle_opts(int argc, char** argv, notcurses_options& opts, bool* quiet,
       case 'V':
         printf("ncplayer version %s\n", notcurses_version());
         exit(EXIT_SUCCESS);
+      case 'n':
+        if(*noninterp){
+          std::cerr <<  "Provided -n twice!" << std::endl;
+          usage(std::cerr, argv[0], EXIT_FAILURE);
+        }
+        *noninterp = true;
+        break;
       case 'a':
         if(*transcolor){
           std::cerr <<  "Provided -a twice!" << std::endl;
@@ -311,7 +320,7 @@ int perframe_direct(struct ncvisual* ncv, struct ncvisual_options* vopts,
 // argc/argv ought already be reduced to only the media arguments
 int direct_mode_player(int argc, char** argv, ncscale_e scalemode,
                        ncblitter_e blitter, int lmargin,
-                       unsigned transcolor){
+                       bool noninterp, unsigned transcolor){
   Direct dm{};
   if(!dm.canopen_images()){
     std::cerr << "Notcurses was compiled without multimedia support\n";
@@ -341,6 +350,9 @@ int direct_mode_player(int argc, char** argv, ncscale_e scalemode,
     vopts.scaling = scalemode;
     vopts.x = static_cast<int>(a);
     vopts.flags = NCVISUAL_OPTION_HORALIGNED;
+    if(noninterp){
+      vopts.flags |= NCVISUAL_OPTION_NOINTERPOLATE;
+    }
     if(transcolor){
       vopts.flags |= NCVISUAL_OPTION_ADDALPHA;
     }
@@ -362,7 +374,7 @@ int rendered_mode_player_inner(NotCurses& nc, int argc, char** argv,
                                ncscale_e scalemode, ncblitter_e blitter,
                                bool quiet, bool loop,
                                double timescale, double displaytime,
-                               uint32_t transcolor){
+                               bool noninterp, uint32_t transcolor){
   int dimy, dimx;
   std::unique_ptr<Plane> stdn(nc.get_stdplane(&dimy, &dimx));
   uint64_t transchan = 0;
@@ -389,6 +401,9 @@ int rendered_mode_player_inner(NotCurses& nc, int argc, char** argv,
     struct ncvisual_options vopts{};
     int r;
     vopts.flags |= NCVISUAL_OPTION_HORALIGNED | NCVISUAL_OPTION_VERALIGNED;
+    if(noninterp){
+      vopts.flags |= NCVISUAL_OPTION_NOINTERPOLATE;
+    }
     if(transcolor){
       vopts.flags |= NCVISUAL_OPTION_ADDALPHA;
     }
@@ -468,7 +483,7 @@ int rendered_mode_player(int argc, char** argv, ncscale_e scalemode,
                          ncblitter_e blitter, notcurses_options& ncopts,
                          bool quiet, bool loop,
                          double timescale, double displaytime,
-                         uint32_t transcolor){
+                         bool noninterp, uint32_t transcolor){
   // no -k, we're using full rendered mode (and the alternate screen).
   ncopts.flags |= NCOPTION_INHIBIT_SETLOCALE;
   if(quiet){
@@ -484,7 +499,7 @@ int rendered_mode_player(int argc, char** argv, ncscale_e scalemode,
     }
     r = rendered_mode_player_inner(nc, argc, argv, scalemode, blitter,
                                    quiet, loop, timescale, displaytime,
-                                   transcolor);
+                                   noninterp, transcolor);
     if(!nc.stop()){
       return -1;
     }
@@ -515,17 +530,18 @@ auto main(int argc, char** argv) -> int {
   uint32_t transcolor = 0;
   bool quiet = false;
   bool loop = false;
+  bool noninterp = false;
   auto nonopt = handle_opts(argc, argv, ncopts, &quiet, &timescale, &scalemode,
-                            &blitter, &displaytime, &loop, &transcolor);
+                            &blitter, &displaytime, &loop, &noninterp, &transcolor);
   int r;
   // if -k was provided, we now use direct mode rather than simply not using the
   // alternate screen, so that output is inline with the shell.
   if(ncopts.flags & NCOPTION_NO_ALTERNATE_SCREEN){
     r = direct_mode_player(argc - nonopt, argv + nonopt, scalemode, blitter,
-                           ncopts.margin_l, transcolor);
+                           ncopts.margin_l, noninterp, transcolor);
   }else{
     r = rendered_mode_player(argc - nonopt, argv + nonopt, scalemode, blitter, ncopts,
-                             quiet, loop, timescale, displaytime, transcolor);
+                             quiet, loop, timescale, displaytime, noninterp, transcolor);
   }
   if(r){
     return EXIT_FAILURE;

--- a/src/player/play.cpp
+++ b/src/player/play.cpp
@@ -372,7 +372,9 @@ int rendered_mode_player_inner(NotCurses& nc, int argc, char** argv,
   struct ncplane_options nopts{};
   // leave a line at the bottom. perhaps one day we'll put information there.
   // for now, this keeps us from scrolling when we use bitmaps.
-  nopts.margin_b = 1;
+  if(nopts.margin_b == 0){
+    nopts.margin_b = 1;
+  }
   nopts.name = "play";
   nopts.resizecb = ncplane_resize_marginalized;
   nopts.flags = NCPLANE_OPTION_MARGINALIZED;

--- a/src/poc/vizdirect.c
+++ b/src/poc/vizdirect.c
@@ -24,7 +24,12 @@ partial_image(struct ncdirect* n, const char* file){
       int cols = x;
       ncdirectv* v;
       printf("Size: %dx%d\n", cols, rows);
-      if((v = ncdirectf_render(n, nf, blit, NCSCALE_NONE, rows, cols)) == NULL){
+      struct ncvisual_options vopts = {
+        .blitter = blit,
+        .leny = rows * geom.scaley,
+        .lenx = cols * geom.scalex,
+      };
+      if((v = ncdirectf_render(n, nf, &vopts)) == NULL){
         ncdirectf_free(nf);
         return -1;
       }

--- a/src/tests/direct.cpp
+++ b/src/tests/direct.cpp
@@ -83,7 +83,9 @@ TEST_CASE("DirectMode") {
     CHECK(475 == geom.pixy);
     CHECK(860 == geom.pixx);
     CHECK(NCBLIT_DEFAULT != blitter);
-    auto ncdv = ncdirectf_render(nc_, dirf, blitter, NCSCALE_NONE, 0, 0);
+    struct ncvisual_options vopts{};
+    vopts.blitter = blitter;
+    auto ncdv = ncdirectf_render(nc_, dirf, &vopts);
     CHECK(nullptr != ncdv);
     CHECK(0 == ncdirect_raster_frame(nc_, ncdv, NCALIGN_LEFT));
     ncdirectf_free(dirf);
@@ -101,7 +103,9 @@ TEST_CASE("DirectMode") {
       CHECK(NCBLIT_PIXEL == blitter);
       CHECK(geom.cdimy == geom.scaley);
       CHECK(geom.cdimx == geom.scalex);
-      auto ncdv = ncdirectf_render(nc_, dirf, blitter, NCSCALE_NONE, 0, 0);
+      struct ncvisual_options vopts{};
+      vopts.blitter = blitter;
+      auto ncdv = ncdirectf_render(nc_, dirf, &vopts);
       CHECK(nullptr != ncdv);
       CHECK(0 == ncdirect_raster_frame(nc_, ncdv, NCALIGN_LEFT));
       ncdirectf_free(dirf);
@@ -112,7 +116,9 @@ TEST_CASE("DirectMode") {
     if(is_test_tty()){
       auto dirf = ncdirectf_from_file(nc_, find_data("worldmap.png").get());
       REQUIRE(nullptr != dirf);
-      auto ncdv = ncdirectf_render(nc_, dirf, NCBLIT_1x1, NCSCALE_NONE, 0, 0);
+      struct ncvisual_options vopts{};
+      vopts.blitter = NCBLIT_1x1;
+      auto ncdv = ncdirectf_render(nc_, dirf, &vopts);
       CHECK(nullptr != ncdv);
       CHECK(0 == ncdirect_raster_frame(nc_, ncdv, NCALIGN_LEFT));
       ncdirectf_free(dirf);
@@ -132,7 +138,10 @@ TEST_CASE("DirectMode") {
       if(ncdirect_check_pixel_support(nc_) > 0){
         auto dirf = ncdirectf_from_file(nc_, find_data("worldmap.png").get());
         REQUIRE(nullptr != dirf);
-        auto ncdv = ncdirectf_render(nc_, dirf, NCBLIT_PIXEL, NCSCALE_NONE, 0, 0);
+        struct ncvisual_options vopts{};
+        vopts.blitter = NCBLIT_PIXEL;
+        vopts.flags = NCVISUAL_OPTION_NODEGRADE;
+        auto ncdv = ncdirectf_render(nc_, dirf, &vopts);
         CHECK(nullptr != ncdv);
         CHECK(0 == ncdirect_raster_frame(nc_, ncdv, NCALIGN_LEFT));
         ncdirectf_free(dirf);

